src/network/networkd-dhcp-static-lease.c

```diff
@@ -0,0 +1,236 @@
+        if (!data)
+                return -ENOMEM;
+
+        /* set client id type to 1: Ethernet Link-Layer (RFC 2132) */
+        data[0] = 0x01;
+        memcpy(data + 1, mac_addr, mac_addr_size);
+
+        free_and_replace(lease->mac_addr, data);
+
+        lease->mac_addr_size = mac_addr_size + 1;
+
+        return 0;
+}
+
+static int dhcp_static_lease_set_address(DHCPStaticLease *lease, const struct in_addr *address) {
+        assert(address);
+        assert(lease);
+        assert_return(address->s_addr != 0, -EINVAL);
+
+        lease->address = (struct in_addr)
+        {
+                .s_addr = address->s_addr,
+        };
+
+        return 0;
+}
+
+static int lease_new_static(Network *network, const char *filename, unsigned section_line, DHCPStaticLease **ret) {
+        _cleanup_(network_config_section_freep) NetworkConfigSection *n = NULL;
+        _cleanup_(dhcp_static_lease_freep) DHCPStaticLease *static_lease = NULL;
+        int r;
+
+        assert(network);
+        assert(ret);
+        assert(filename);
+        assert(section_line > 0);
+        
+        r = network_config_section_new(filename, section_line, &n);
+        if (r < 0)
+                return r;
+
+        static_lease = ordered_hashmap_get(network->dhcp_static_leases_by_section, n);
+        if (static_lease) {
+                *ret = TAKE_PTR(static_lease);
+                return 0;
+        }
+
+        r = dhcp_static_lease_new(&static_lease);
+        if (r < 0)
+                return r;
+
+        static_lease->network = network;
+
+        static_lease->section = TAKE_PTR(n);
+        r = ordered_hashmap_ensure_put(&network->dhcp_static_leases_by_section, &network_config_hash_ops, static_lease->section, static_lease);
+        if (r < 0)
+                return r;
+
+        *ret = TAKE_PTR(static_lease);
+
+        return 0;
+}
+
+void network_drop_invalid_static_leases(Network *network) {
+        DHCPStaticLease *static_lease;
+
+        assert(network);
+
+        ORDERED_HASHMAP_FOREACH(static_lease, network->dhcp_static_leases_by_section) {
+                assert(static_lease);
+
+                if (section_is_invalid(static_lease->section))
+                        dhcp_static_lease_free(static_lease);
+        }
+}
+
+int config_parse_dhcp_static_lease_address(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        Network *network;
+        _cleanup_(dhcp_static_lease_free_or_set_invalidp) DHCPStaticLease *current_lease = NULL;
+        union in_addr_union addr;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(userdata);
+
+        network = userdata;
```

Comment:
You can merge this with the variable declaration in the above: `Network *network = userdata;`.